api/agent/pure_runner.go

```diff
@@ -27,171 +28,356 @@ import (
 	"google.golang.org/grpc/peer"
 )
 
+var (
+	ErrorExpectedTry  = errors.New("Protocol failure: expected ClientMsg_Try")
+	ErrorExpectedData = errors.New("Protocol failure: expected ClientMsg_Data")
+)
+
 // callHandle represents the state of the call as handled by the pure runner, and additionally it implements the
 // interface of http.ResponseWriter so that it can be used for streaming the output back.
 type callHandle struct {
 	engagement runner.RunnerProtocol_EngageServer
+	ctx        context.Context
 	c          *call // the agent's version of call
-	input      io.WriteCloser
-	started    bool
-	done       chan error // to synchronize
-	// As the state can be set and checked by both goroutines handling this state, we need a mutex.
-	stateMutex sync.Mutex
+
 	// Timings, for metrics:
 	receivedTime  strfmt.DateTime // When was the call received?
 	allocatedTime strfmt.DateTime // When did we finish allocating capacity?
-	// Last communication error on the stream (if any). This basically acts as a cancellation flag too.
-	streamError error
+
 	// For implementing http.ResponseWriter:
-	outHeaders    http.Header
-	outStatus     int
-	headerWritten bool
+	headers http.Header
+	status  int
+
+	headerOnce        sync.Once
+	shutOnce          sync.Once
+	pipeToFnCloseOnce sync.Once
+
+	outQueue  chan *runner.RunnerMsg
+	doneQueue chan struct{}
+	errQueue  chan error
+	inQueue   chan *runner.ClientMsg
+	pipeToFn  io.WriteCloser
+}
+
+func NewCallHandle(engagement runner.RunnerProtocol_EngageServer) *callHandle {
+	state := &callHandle{
+		engagement: engagement,
+		ctx:        engagement.Context(),
+		headers:    make(http.Header),
+		status:     200,
+		outQueue:   make(chan *runner.RunnerMsg),
+		doneQueue:  make(chan struct{}),
+		errQueue:   make(chan error, 1), // always allow one error (buffered)
+		inQueue:    make(chan *runner.ClientMsg),
+	}
+
+	// spawn receiver and sender go-routines. We can work
+	// concurrently on engagement Send()/Recv() separately, but multiple
+	// go-routines cannot issue Send() at the same time.
```

Comment:
Based upon our offline discussion, it might be worth beefing this comment up to explain that the limitation comes from GRPC.  Multiple go routines may neither send on the same stream, nor receive on the same stream.  However, it is permissible to divide send and receive as long as only one go routine handles each.  It would be nice if we had some kind of invariant available to enforce this, too.

Reply:
added some comments and link to the thread-safety documentation in gRPC.